util/network/src/client_version.rs

```diff
@@ -0,0 +1,473 @@
+// need to be abstracted away into a trait.
+impl ParityClientData {
+	fn name(&self) -> &str {
+		self.name.as_str()
+	}
+
+	fn variant(&self) -> Option<&str> {
+		self.variant.as_ref().map(String::as_str)
+	}
+
+	fn semver(&self) -> &Version {
+		&self.semver
+	}
+
+	fn os(&self) -> &str {
+		self.os.as_str()
+	}
+
+	fn compiler(&self) -> &str {
+		self.compiler.as_str()
+	}
+}
+
+#[derive(Clone,Debug,PartialEq,Eq)]
+pub enum ClientVersion {
+	ParityClient(
+		ParityClientData
+	),
+	ParityUnknownFormat(String),
+	Other(String), // Id string
+}
+
+// TODO: Maybe merge with Peercapabilityinfo in ethcore-network?
+pub trait ClientCapabilities {
+	fn can_handle_large_requests(&self) -> bool;
+
+	fn accepts_service_transaction(&self) -> bool;
+}
+
+// This is an implementation of a function taken from propagator.rs
+fn parity_accepts_service_transaction(parity_client_data: &ParityClientData) -> bool {
+	let service_transactions_version = Version::parse(SERVICE_TRANSACTIONS_VERSION).unwrap();
+
+	*parity_client_data.semver() >= service_transactions_version
+}
+
+impl ClientCapabilities for ClientVersion {
+	fn can_handle_large_requests(&self) -> bool {
+		match self {
+			ClientVersion::ParityClient(data) => {
+				if *data.semver() < Version::parse(PARITY_CLIENT_LARGE_REQUESTS_VERSION).unwrap() {
+					false
+				} else {
+					true
+				}
+			},
+			ClientVersion::ParityUnknownFormat(_) => false, // Play it safe
+			ClientVersion::Other(_) => true // As far as we know
+		}
+	}
+
+	/// Checks if peer is able to process service transactions
+	fn accepts_service_transaction(&self) -> bool {
+		match self {
+			ClientVersion::ParityClient(data) => parity_accepts_service_transaction(&data),
+			ClientVersion::ParityUnknownFormat(_) => false,
+			ClientVersion::Other(_) => false
+		}
+	}
+
+}
+
+
+fn is_parity(client_id: &str) -> bool {
+	if client_id.starts_with(LEGACY_CLIENT_ID_PREFIX) || client_id.starts_with(PARITY_CLIENT_ID_PREFIX) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+// Parse known parity formats.
+//
+// This is really not robust: parse four arguments and
+// allow for an extra argument between identifier and
+// version
+// TODO implement a better logic
+fn parse_parity_format(client_version: &str) -> Result<ParityClientData, ()> {
+	let tokens: Vec<&str> = client_version.split("/").collect();
+
+
+	// Basically strip leading 'v'
+	if let Some(version_number) = &get_number_from_version(tokens[1]) {
+		if tokens.len() < 4 {
+			return Err(());
+		}
+
+		return Ok(
+			ParityClientData {
+				name: tokens[0].to_string(),
+				variant: None,
+				semver: Version::parse(version_number).unwrap(),
+				os: tokens[2].to_string(),
+				compiler: tokens[3].to_string(),
+			}
+		);
+	} else if let Some(version_number) = &get_number_from_version(tokens[2]) {
+		if tokens.len() < 5 {
+			return Err(());
+		}
+
+		return Ok(
+			ParityClientData {
+				name: tokens[0].to_string(),
+				variant: Some(tokens[1].to_string()),
```

Comment:
You are right here. We have two options: we stick to the two basic alternatives we have right now, namely a four and a five argument ID strings with versions in second and third position respectively, or we go to a more generic solution that tries to locate version-like tokens wherever they are. I would go with the simplest one right now.

Reply:
I think [finding the rightmost `/v`](https://github.com/paritytech/parity-ethereum/pull/10247/files#diff-68aa829147cff9109153923db5442c70L53) was good enough generic solution, alternatively, we can disallow having `/` in `--identity` and go for the simplest one.

Reply:
How safe is it to assume version is the third token from the right? Os/target/platform and compiler are always there in parity IDs, right? Because if not, say the string is "Parity/linux/rustc/v2.3.0". Is "linux/rustc" platform and compiler or identity?

For parity, I would expect, from right to left: compiler, os, version, then the string should begin with "Parity/" or "Parity-Ethereum/" and everything in between is identity.

Reply:
> Os/target/platform and compiler are always there in parity IDs, right?

According to git history of [util/version](https://github.com/paritytech/parity-ethereum/blob/eb3d33ed6ff2a697e9c48eb025c044e6b872f468/util/version/src/lib.rs#L52-L57), yes, at least for the versions we care about.

> Is "linux/rustc" platform and compiler or identity?

Identity (if present) is always inserted before version, so it would be identity (though the string is an invalid version).

> For parity, I would expect, from right to left: compiler, os, version, then the string should begin with "Parity/" or "Parity-Ethereum/" and everything in between is identity.

sounds right

Reply:
Then we have an agreement :-)